### PR TITLE
Calculate tour costs

### PR DESCRIPTION
**Pull request by atte-wsp**
* Generalized and out-of-pocket costs of tours based on choices

### CODE REVIEW

File: Scripts/datatypes/tour.py [hunk 1]
```diff
@@ -179,6 +181,24 @@ def choose_secondary_destination(self, cumulative_probs):
                     + numpy.searchsorted(cumulative_probs, self._sec_dest_draw))
         self.position = (self.position[0], self.position[1], dest_idx)
         self.purpose.sec_dest_purpose.attracted_tours[self.mode][dest_idx] += 1
+    
+    def calc_cost(self):
+        """Construct cost and time components from tour dest choice. """
+        impedance = self.purpose.model.impedance[self.mode]
+        time = self._get_cost(impedance["time"])
+        self.cost = self._get_cost(impedance["cost"])
```
Comment (zptro) [inline Scripts/datatypes/tour.py:189]: Will this not cause a `KeyError` for bike and walk?

**Review by zptro — COMMENTED** [review]
Nice attempt! The secondary destinations are not handled exactly right (see comment below) and memory consumption will be high. I am thinking, should we have a simpler approach using the un-transposed impedance matrices (e.g., assuming aht-iht pattern for all work tours) and adding tour legs one-by-one?

File: Scripts/datatypes/tour.py [hunk 2]
```diff
@@ -179,6 +181,24 @@ def choose_secondary_destination(self, cumulative_probs):
...
+        self.gen_cost = self.cost + time * vot[self.purpose.name]
+
+    def _get_cost(self, mtx):
+        """Check if matrix and return value from position. """
+        cost = 0
+        if isinstance(mtx, numpy.ndarray):
```
[the remark this answers is not in the document]
Comment (atte-wsp) [reply Scripts/datatypes/tour.py:195]: For walk and bike `mtx` is sometimes float zero and there would be index error. Probably this should be handled with try-except clause?

File: Scripts/datatypes/tour.py (hunk 1 above)
Comment (atte-wsp) [reply Scripts/datatypes/tour.py:189]: I think this is related to above. Bike and walk return float zeros.

File: Scripts/parameters/impedance_transformation.py [hunk 3]
```diff
@@ -275,3 +275,17 @@
...
+    "hw": 0.132, # eur / min
+    "hu": 0.132,
+    "hc": 0.132,
+    "hs": 0.0895,
+    "ho": 0.0895,
+    "wo": 0.0895,
+    "woo": 0.0895,
+    "oo": 0.0895,
+    "wh": 0.0895,
+}
```
[the remark this answers is not in the document]
Comment (atte-wsp) [reply Scripts/parameters/impedance_transformation.py:291]: True, I made this as a quick first draft and forgot to fix it later.

File: Scripts/models/logit.py [hunk 4]
```diff
@@ -318,6 +318,7 @@ def calc_basic_prob(self, impedance):
             Mode (car/transit/bike/walk) : numpy 2-d matrix
                 Choice probabilities
         """
+        self.impedance = impedance
```
[the remark this answers is not in the document]
Comment (atte-wsp) [reply Scripts/models/logit.py:321]: True, after monitoring the memory consumption this does need some changes.  Didn't realize that impact on memory consumption is that big. Probably refactoring with simpler solution is the best option as secondary destination require changes as well.

**Comment by atte-wsp** [comment]
@zptro Made new version with simplified approach. Would this be more efficient? Not sure if I'm completely happy with this implementation as the cost calculation has quite many exceptions, but could not figure out better solutions at this point.

Results are reasonable and highlight more differences between Helsinki cbd and other region than the previous implementation.

File: Scripts/datatypes/tour.py [hunk 5]
```diff
@@ -179,6 +182,54 @@ def choose_secondary_destination(self, cumulative_probs):
...
+            # first leg of tour
+            cost += departure_imp[self.position[0], self.position[1]]
+            # check if tour has secondary destination and add accordingly
+            if len(self.position) > 2:
+                cost += sec_dest_imp[self.position[1], self.position[2]]
+                cost += return_imp[self.position[2], self.position[0]]
+            else: 
+                cost += return_imp[self.position[1], self.position[0]]
+        except KeyError:
+            pass
```
Comment (zptro) [inline Scripts/datatypes/tour.py:227]: I think this try-except clause could be considered to wide, we risk excepting unintended bugs. But I cannot come up with a more elegant solution.

File: Scripts/transform/impedance_transformer.py [hunk 6]
```diff
@@ -47,11 +47,9 @@ def transform(self, purpose, impedance):
...
-                day_imp["transit"]["cost"], trips_month["transit_work"][0])
-            trips_per_month[purpose.ubounds, :] = trips_month["transit_work"][1]
-            day_imp["transit"]["cost"] /= trips_per_month
-        else:
-            day_imp["transit"]["cost"] /= trips_month["transit_leisure"]
+        transit_class = "{}_{}".format("transit", assignment_classes[purpose.name])
+        trips_per_month = numpy.full_like(
+            day_imp["transit"]["cost"], trips_month[transit_class][0])
+        trips_per_month[purpose.ubounds, :] = trips_month[transit_class][1]
+        day_imp["transit"]["cost"] /= trips_per_month
```
Comment (zptro) [inline Scripts/transform/impedance_transformer.py:54]: This generalization is nice!

**Review by zptro — APPROVED** [review]
Yes, I think this should be a working and stable solution. Well done!

File: Scripts/datatypes/tour.py (hunk 5 above)
Comment (zptro) [reply Scripts/datatypes/tour.py:227]: Maybe we could add a comment describing what the except clause is supposed to catch?

**Pull request merged**